x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/slm/SnapshotRetentionConfiguration.java

```diff
@@ -144,68 +146,68 @@ public Integer getMaximumSnapshotCount() {
-                                "{} non-failed snapshots ({} total), over the limit of {} maximum snapshots",
-                            snapName, successfulSnapsToDelete, successfulSnapshotCount, totalSnapshotCount, this.maximumSnapshotCount);
-                        return true;
-                    } else {
-                        logger.trace("[{}]: SKIPPING as it is not one of the {} oldest snapshots with " +
-                                "{} non-failed snapshots ({} total), over the limit of {} maximum snapshots",
-                            snapName, successfulSnapsToDelete, successfulSnapshotCount, totalSnapshotCount, this.maximumSnapshotCount);
+            if (this.maximumSnapshotCount != null && successfulSnapshotCount > this.maximumSnapshotCount) {
+                final long successfulSnapsToDelete = successfulSnapshotCount - this.maximumSnapshotCount;
+                boolean found = false;
+                int successfulSeen = 0;
+                for (SnapshotInfo s : sortedSnapshots) {
+                    if (s.state() == SnapshotState.SUCCESS) {
+                        successfulSeen++;
                     }
+                    if (successfulSeen > successfulSnapsToDelete) {
+                        break;
+                    }
+                    if (s.equals(si)) {
+                        found = true;
+                        break;
+                    }
+                }
+                if (found) {
+                    logger.trace("[{}]: ELIGIBLE as it is one of the {} oldest snapshots with " +
+                        "{} non-failed snapshots ({} total), over the limit of {} maximum snapshots",
+                        snapName, successfulSnapsToDelete, successfulSnapshotCount, totalSnapshotCount, this.maximumSnapshotCount
+                    );
+                    return true;
+                } else {
+                    logger.trace("[{}]: SKIPPING as it is not one of the {} oldest snapshots with " +
+                        "{} non-failed snapshots ({} total), over the limit of {} maximum snapshots",
+                        snapName, successfulSnapsToDelete, successfulSnapshotCount, totalSnapshotCount, this.maximumSnapshotCount
+                    );
                 }
             }
 
             // Next check the minimum count, since that is a blanket requirement regardless of time,
             // if we haven't hit the minimum then we need to keep the snapshot regardless of
             // expiration time
-            if (this.minimumSnapshotCount != null) {
-                if (successfulSnapshotCount <= this.minimumSnapshotCount)
-                    if (unsuccessfulStates.contains(si.state()) == false) {
-                        logger.trace("[{}]: INELIGIBLE as there are {} non-failed snapshots ({} total) and {} minimum snapshots needed",
+            if (this.minimumSnapshotCount != null && successfulSnapshotCount <= this.minimumSnapshotCount) {
```

Comment:
I did originally put these on separate lines so that visually the `minimumSnapshotCount` check indicated whether that constraint was present (same with the other checks). It's fine to collapse them, just figured I would explain why they were originally like that (mostly personal preference and code aestheticism).